docs/stress-test.md

```diff
@@ -122,3 +141,128 @@ platform is deployment-specific but the following are some examples of dashboard
+	  0.00%   0.00%   0.010s    0.010s   add_timeout (tornado/ioloop.py:580)
+	  0.00%   0.00%   0.000s    0.010s   <module> (jupyterhub:11)
+	  0.00%   0.00%   0.000s    0.010s   start (tornado/platform/asyncio.py:149)
+	  0.00%   0.00%   0.000s    0.010s   _run_callback (tornado/ioloop.py:743)
+	  0.00%   0.00%   0.000s    0.010s   _schedule_next (tornado/ioloop.py:916)
+	  0.00%   0.00%   0.000s    0.010s   launch_instance (jupyterhub/app.py:2782)
+	  0.00%   0.00%   0.000s    0.010s   run_forever (asyncio/base_events.py:438)
+	  0.00%   0.00%   0.000s    0.010s   _run (tornado/ioloop.py:911)
+	  0.00%   0.00%   0.000s    0.010s   _run_once (asyncio/base_events.py:1451)
+	  0.00%   0.00%   0.000s    0.010s   _run (asyncio/events.py:145)
+    ```
+    The output can be sorted by each column as well.
+<a name="py-spy-record"></a>
+1. `py-spy record`  
+The record command runs in the foreground collecting samples and when it's closed (either
+by CTRL+C or when it reaches its configured duration) an SVG flamegraph is written to disk.
+    ```
+    jovyan@hub-69f94ddc84-g26qj:/$ py-spy record -o /tmp/py-spy-trace -p 1
+    py-spy> Sampling process 100 times a second. Press Control-C to exit.
+
+    ^C
+    py-spy> Stopped sampling because Control-C pressed
+    py-spy> Wrote flamegraph data to '/tmp/py-spy-trace'. Samples: 20541 Errors: 0
+    ```
+    That will produce an SVG like this:
+    ![Example](./images/py-spy-example.svg)
+<a name="py-spy-dump"></a>
+1. `py-spy dump`  
+The dump command will dump the state of all threads for the specified process. It can
+optionally show the local variables for each frame. This is helpful for figuring out why
+the hub process may appear stuck for example.
+
+	```
+	jovyan@hub-69f94ddc84-g26qj:/$ py-spy dump -p 1 --locals
+	Process 1: /usr/bin/python3 /usr/local/bin/jupyterhub --config /etc/jupyterhub/jupyterhub_config.py --upgrade-db
+	Python v3.6.9 (/usr/bin/python3.6)
+
+	Thread 1 (idle): "MainThread"
+		select (selectors.py:445)
+			Arguments::
+				self: <EpollSelector at 0x7f79ac4ab898>
+				timeout: 0.998
+			Locals::
+				max_ev: 3
+				ready: []
+		_run_once (asyncio/base_events.py:1415)
+			Arguments::
+				self: <_UnixSelectorEventLoop at 0x7f79ac4ab7b8>
+			Locals::
+				sched_count: 45
+				timeout: 0.9979082886129618
+				when: 5094106.57817052
+		run_forever (asyncio/base_events.py:438)
+			Arguments::
+				self: <_UnixSelectorEventLoop at 0x7f79ac4ab7b8>
+			Locals::
+				old_agen_hooks: (None, None)
+		start (tornado/platform/asyncio.py:149)
+			Arguments::
+				self: <AsyncIOMainLoop at 0x7f79ac447be0>
+			Locals::
+				old_loop: <_UnixSelectorEventLoop at 0x7f79ac4ab7b8>
+		launch_instance (jupyterhub/app.py:2782)
+			Arguments::
+				cls: <MetaHasTraits at 0x30ee728>
+				argv: None
+			Locals::
+				self: <JupyterHub at 0x7f79b12b7358>
+				loop: <AsyncIOMainLoop at 0x7f79ac447be0>
+				task: <_asyncio.Task at 0x7f79abd788c8>
+		<module> (jupyterhub:11)
+	```
```

Comment:
Should we have a `### Troubleshooting` section as well about anything you hit while figuring out how to use py-spy? The thing that comes to mind is when you'd get a random pause which I thought was related to this https://github.com/benfred/py-spy#how-can-i-avoid-pausing-the-python-program but I'm not sure.

Reply:
I was never able to track down why the process wouldn't resume sometimes. I'm pretty sure passing `--nonblocking` would prevent that but when I tried it that way the sampling error rate went way up.